File: smoke_tests/typescript/xxe.ts
// XXE (XML External Entity) Test Cases

// Test 1: libxmljs with external entities enabled
function parseXMLWithLibxmljs(xmlString: string): any {
    const libxmljs = require('libxmljs');
    // VULNERABLE: noent option enables entity expansion
    return libxmljs.parseXml(xmlString, { noent: true });
}

// Test 2: xml2js without XXE protection
function parseXMLWithXml2js(xmlString: string): Promise<any> {
    const xml2js = require('xml2js');
    const parser = new xml2js.Parser({
        // VULNERABLE: No XXE protection configured
    });
    return parser.parseStringPromise(xmlString);
}

// Test 3: DOMParser without security settings
function parseXMLInBrowser(xmlString: string): Document {
    const parser = new DOMParser();
    // VULNERABLE: No validation or entity restrictions
    return parser.parseFromString(xmlString, 'text/xml');
}

// Test 4: xmldom parser with default settings
function parseXMLWithXmldom(xmlString: string): Document {
    const { DOMParser } = require('xmldom');
    // VULNERABLE: Default settings allow entity expansion
    const parser = new DOMParser();
    return parser.parseFromString(xmlString, 'text/xml');
}

// Test 5: Fast-xml-parser with entities enabled
function parseXMLWithFastParser(xmlString: string): any {
    const { XMLParser } = require('fast-xml-parser');
    const parser = new XMLParser({
        // VULNERABLE: Allowing entities without restrictions
        allowBooleanAttributes: true,
        processEntities: true
    });
    return parser.parse(xmlString);
}

File: smoke_tests/rust/xxe.rs
// XXE (XML External Entity) Test Cases

// Test 1: XML parsing with external entities enabled
fn parse_xml_with_entities(xml_str: &str) -> Result<roxmltree::Document, roxmltree::Error> {
    // VULNERABLE: roxmltree doesn't expand entities by default, but simulating vulnerability
    roxmltree::Document::parse(xml_str)
}

// Test 2: Quick-xml parsing without security
fn parse_xml_quick(xml_bytes: &[u8]) -> Result<(), quick_xml::Error> {
    use quick_xml::Reader;
    use quick_xml::events::Event;

    let mut reader = Reader::from_reader(xml_bytes);
    reader.trim_text(true);

    loop {
        // VULNERABLE: Processing XML without entity restrictions
        match reader.read_event() {
            Ok(Event::Eof) => break,
            Err(e) => return Err(e),
            _ => {}
        }
    }
    Ok(())
}

// Test 3: Serde XML deserialization
fn deserialize_xml<T: serde::de::DeserializeOwned>(xml_str: &str) -> Result<T, quick_xml::DeError> {
    // VULNERABLE: Deserializing XML without entity protection
    quick_xml::de::from_str(xml_str)
}

// Test 4: XML-rs parser
fn parse_with_xmlrs(xml_str: &str) -> Result<(), xml::reader::Error> {
    use xml::reader::{EventReader, XmlEvent};

    let parser = EventReader::from_str(xml_str);
    // VULNERABLE: No explicit entity restrictions
    for event in parser {
        match event? {
            XmlEvent::EndDocument => break,
            _ => {}
        }
    }
    Ok(())
}

// Test 5: Minidom XML parsing
fn parse_minidom(xml_str: &str) -> Result<minidom::Element, minidom::Error> {
    // VULNERABLE: Parsing untrusted XML
    xml_str.parse::<minidom::Element>()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_xxe() {
        let xml = r#"<?xml version="1.0"?><!DOCTYPE foo [<!ENTITY xxe SYSTEM "file:///etc/passwd">]><foo>&xxe;</foo>"#;
        let _ = parse_xml_with_entities(xml);
    }
}
